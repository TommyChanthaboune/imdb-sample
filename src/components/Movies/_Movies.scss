.movies {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  min-height: 852px;
  .poster {
    margin: 0 1rem 1rem 0;
  }
}

@include xlarge-screens {
  .movies {
    justify-content: flex-start;

    .poster {
      flex-basis: calc(20% - 1rem);
    }
  }
}
