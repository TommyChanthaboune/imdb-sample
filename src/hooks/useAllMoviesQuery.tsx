import { useQuery } from '@apollo/react-hooks';
import { gql, ApolloError } from '@apollo/client';
import { IMovie } from '../interfaces';

/**
 * NOTE: Due to limitations with JSON-server,
 * this query does not fully reflect each field and
 * associated type, but will return the full set of data
 * regardless.
 * */

const ALL_MOVIES_QUERY = gql`
  query allMovies {
    allMovies {
      id
      title
      voteCount
      voteAverage
      posterPath
      genres
    }
  }
`;

export const useAllMoviesQuery = (): {
  data: IMovie[];
  error: ApolloError | undefined;
  loading: Boolean;
} => {
  const { data, loading, error } = useQuery(ALL_MOVIES_QUERY, {
    fetchPolicy: 'cache-and-network'
  });

  const result = data?.allMovies ?? [];

  return {
    data: result,
    error,
    loading
  };
};
