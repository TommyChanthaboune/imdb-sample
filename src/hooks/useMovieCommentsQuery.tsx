/* eslint-disable @typescript-eslint/no-unused-vars */
import { useEffect } from 'react';
import { useQuery, useMutation } from '@apollo/react-hooks';
import { gql, ApolloError } from '@apollo/client';
import { IComment, IComments, IAddComment, ICommentMeta } from '../interfaces';

interface IMovieCommentsQueryReturn {
  commentData: IComment[];
  commentError: ApolloError | undefined;
  commentLoading: Boolean;
  commentCount: number | undefined;
  commentRefetch: any;
  addComment: any;
}

const COMMENT_QUERY = gql`
  query allComments($movie_id: ID) {
    allComments(filter: { movie_id: $movie_id }, sortOrder: "dsc") {
      id
      title
      author
      data
    }
  }
`;

const COMMENT_META = gql`
  query _allCommentsMeta($movie_id: ID) {
    _allCommentsMeta(filter: { movie_id: $movie_id }) {
      count
    }
  }
`;

const ADD_COMMENT = gql`
  mutation createComment(
    $movie_id: ID!
    $author: String!
    $title: String!
    $data: String!
  ) {
    createComment(
      movie_id: $movie_id
      author: $author
      title: $title
      data: $data
    ) {
      id
    }
  }
`;

export const useMovieCommentsQuery = (
  movie_id: string
): IMovieCommentsQueryReturn => {
  const comment = useQuery<IComments>(COMMENT_QUERY, {
    fetchPolicy: 'cache-and-network',
    variables: {
      movie_id
    }
  });

  const commentMeta = useQuery<ICommentMeta>(COMMENT_META, {
    fetchPolicy: 'cache-and-network',
    variables: {
      movie_id
    }
  });

  const [addComment, { data }] = useMutation<IAddComment, IComment>(
    ADD_COMMENT
  );

  const comments = comment.data?.allComments ?? [];

  return {
    commentData: comments,
    commentError: comment.error,
    commentLoading: comment.loading,
    commentRefetch: comment.refetch,
    // eslint-disable-next-line no-underscore-dangle
    commentCount: commentMeta.data?._allCommentsMeta.count,
    addComment
  };
};
