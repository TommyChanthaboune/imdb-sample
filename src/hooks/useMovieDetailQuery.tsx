/* eslint-disable @typescript-eslint/no-unused-vars */
import { useQuery } from '@apollo/react-hooks';
import { gql, ApolloError } from '@apollo/client';
import { IMovieDetail } from '../interfaces';

interface IMovieDetailQueryReturn {
  data: IMovieDetail;
  error: ApolloError | undefined;
  loading: Boolean;
}

const MOVIE_QUERY = gql`
  query Movie($id: ID!) {
    Movie(id: $id) {
      id
      originalLanguage
      originalTitle
      overview
      popularity
      posterPath
      releaseDate
      title
      voteAverage
      voteCount
      budget
      runtime
      genres
      cast
      director
    }
  }
`;

export const useMovieDetailQuery = (id: string): IMovieDetailQueryReturn => {
  const { data, loading, error } = useQuery(MOVIE_QUERY, {
    fetchPolicy: 'cache-and-network',
    variables: {
      id
    }
  });

  const movie = data?.Movie ?? {};

  return {
    data: movie,
    error,
    loading
  };
};
