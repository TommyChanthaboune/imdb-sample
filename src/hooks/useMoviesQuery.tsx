/* eslint-disable @typescript-eslint/no-unused-vars */
import { useQuery } from '@apollo/react-hooks';
import { gql, ApolloError } from '@apollo/client';
import { IMovieSort, IPaging, IFilterConditons } from '../interfaces';
import { SortOptions } from '../enums';

export interface IMovieQueryReturn {
  data: IMovieSort[];
  error: ApolloError | undefined;
  loading: Boolean;
}

const MOVIE_QUERY = gql`
  query allMovies($sortField: String, $page: Int, $filter: MovieFilter) {
    allMovies(
      sortField: $sortField
      sortOrder: "dsc"
      page: $page
      filter: $filter
    ) {
      id
      title
      voteCount
      voteAverage
      posterPath
      genres
      popularity
      budget
      runtime
      releaseDate
    }
  }
`;

export const useMoviesQuery = (
  paging: IPaging,
  sortOptions: SortOptions,
  filterOptions?: IFilterConditons
): IMovieQueryReturn => {
  const { data, loading, error } = useQuery(MOVIE_QUERY, {
    fetchPolicy: 'cache-and-network',
    variables: {
      sortField: SortOptions[sortOptions],
      filter: filterOptions,
      ...paging
    }
  });

  const result = data?.allMovies ?? [];

  return {
    data: result,
    error,
    loading
  };
};
