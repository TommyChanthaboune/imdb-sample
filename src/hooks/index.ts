import { useAllMoviesQuery } from './useAllMoviesQuery';
import { useMoviesQuery } from './useMoviesQuery';
import { useMovieDetailQuery } from './useMovieDetailQuery';
import { useMovieCommentsQuery } from './useMovieCommentsQuery';

export {
  useAllMoviesQuery,
  useMoviesQuery,
  useMovieDetailQuery,
  useMovieCommentsQuery
};
