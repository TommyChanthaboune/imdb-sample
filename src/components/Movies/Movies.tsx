import React, { FC } from 'react';
import { Poster } from '../Poster/Poster';
import { IPaging, IFilterConditons } from '../../interfaces';
import { SortOptions } from '../../enums/index';
import { useMoviesQuery } from '../../hooks';

interface IMovies {
  paging: IPaging;
  sortBy?: SortOptions;
  filterBy?: IFilterConditons;
}

export const Movies: FC<IMovies> = ({
  paging,
  sortBy = SortOptions.popularity,
  filterBy = {}
}) => {
  const { data, loading } = useMoviesQuery(paging, sortBy, filterBy);

  return (
    <div className="movies">
      {loading ? (
        <p> Loading... </p>
      ) : (
        <>
          {data.length ? (
            data.map((movie) => (
              <Poster
                key={movie.id}
                source={movie.posterPath}
                altText={movie.title}
                id={movie.id}
              />
            ))
          ) : (
            <div>
              <h3>There are no results for your query.</h3>
            </div>
          )}
        </>
      )}
    </div>
  );
};
