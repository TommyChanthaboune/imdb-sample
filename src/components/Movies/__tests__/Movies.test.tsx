import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Movies } from '../Movies';
import mockData from './mockdata.json';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { useMoviesQuery } from '../../../hooks/useMoviesQuery';

jest.mock('../../../hooks/useMoviesQuery.tsx', () => ({
  useMoviesQuery: () => ({
    data: mockData,
    loading: false
  })
}));

describe('Movie Page', () => {
  it('should render correctly', () => {
    const component = render(
      <Router>
        <Movies paging={{ page: 1 }} />
      </Router>
    );

    expect(component).toMatchSnapshot();
  });
});
