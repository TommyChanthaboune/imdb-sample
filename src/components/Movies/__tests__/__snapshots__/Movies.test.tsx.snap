// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Page should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="movies"
      >
        <div
          class="poster"
        >
          <a
            href="/detail?movie=384018"
          >
            <img
              alt="Fast & Furious Presents: Hobbs & Shaw"
              src="https://image.tmdb.org/t/p/w300_and_h450_bestv2//qRyy2UmjC5ur9bDi3kpNNRCc5nc.jpg"
            />
          </a>
        </div>
        <div
          class="poster"
        >
          <a
            href="/detail?movie=582306"
          >
            <img
              alt="Assassin 33 A.D."
              src="https://image.tmdb.org/t/p/w300_and_h450_bestv2//8jDvtdH327I8TgX3UPdkAsZF1dA.jpg"
            />
          </a>
        </div>
        <div
          class="poster"
        >
          <a
            href="/detail?movie=777350"
          >
            <img
              alt="Dory's Reef Cam"
              src="https://image.tmdb.org/t/p/w300_and_h450_bestv2//mMWLGu9pFymqipN8yvISHsAaj72.jpg"
            />
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="movies"
    >
      <div
        class="poster"
      >
        <a
          href="/detail?movie=384018"
        >
          <img
            alt="Fast & Furious Presents: Hobbs & Shaw"
            src="https://image.tmdb.org/t/p/w300_and_h450_bestv2//qRyy2UmjC5ur9bDi3kpNNRCc5nc.jpg"
          />
        </a>
      </div>
      <div
        class="poster"
      >
        <a
          href="/detail?movie=582306"
        >
          <img
            alt="Assassin 33 A.D."
            src="https://image.tmdb.org/t/p/w300_and_h450_bestv2//8jDvtdH327I8TgX3UPdkAsZF1dA.jpg"
          />
        </a>
      </div>
      <div
        class="poster"
      >
        <a
          href="/detail?movie=777350"
        >
          <img
            alt="Dory's Reef Cam"
            src="https://image.tmdb.org/t/p/w300_and_h450_bestv2//mMWLGu9pFymqipN8yvISHsAaj72.jpg"
          />
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
